### PR TITLE
Rule-based 3D renderer

### PR DESCRIPTION
**Pull request by wonder-sk**
In the usual last-minute pre-feature-freeze maniacal effort, here's the rule-based 3D renderer for vector layers. Until now all features of a vector layer had to use the same 3D renderer, but now just like with 2D rendering or labeling it is possible to define hierarchical rules on how to render vector layer in 3D view.

rulebased-3d

One major difference between 2D and 3D rule-based rendering is that rules do not include scale range: while in 2D rendering we always know the scale, in 3D rendering we need to evaluate rules before scale is known (and it is variable in the 3D view anyway).

Includes internal refactoring 3D symbol implementations... before, each 3D symbol would have its own feature loop (actually two: one for non-selected and one for selected features). Now 3D symbols are happy to be a part of an external feature loop thanks to the introduction of QgsFeature3DHandler. There's also a new class Qgs3DRenderContext which is similar to the usual 2D rendering's QgsRenderContext.

TODO: adapt 3D point symbol implementation to use QgsFeature3DHandler

### CODE REVIEW

File: src/3d/qgsfeature3dhandler_p.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Qgs3DSymbolImpl
+{
+
+  Qt3DCore::QEntity *entityFromHandler( QgsFeature3DHandler *handler, const Qgs3DMapSettings &map, QgsVectorLayer *layer )
+  {
+    Qgs3DRenderContext context( map );
+
+    QgsExpressionContext exprContext;
+    exprContext << QgsExpressionContextUtils::globalScope()
```
Comment (nyalldawson) [inline src/3d/qgsfeature3dhandler_p.cpp]: layer scope too?

**Comment by nyalldawson** [comment]
Neat! Will the same restructuring also allow data defined 3d symbol colors in future too?

File: src/3d/qgs3dutils.h [hunk 2]
```diff
@@ -99,6 +100,8 @@ class _3D_EXPORT Qgs3DUtils
     //! Transforms a world point from (origin1, crs1) to (origin2, crs2)
     static QgsVector3D transformWorldCoordinates( const QgsVector3D &worldPoint1, const QgsVector3D &origin1, const QgsCoordinateReferenceSystem &crs1, const QgsVector3D &origin2, const QgsCoordinateReferenceSystem &crs2,
         const QgsCoordinateTransformContext &context );
+
+    static QgsAbstract3DSymbol *symbolForGeometryType( QgsWkbTypes::GeometryType geomType );
```
Comment (nyalldawson) [inline src/3d/qgs3dutils.h]: This should return a std::unique_ptr

File: src/3d/qgsfeature3dhandler_p.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                << QgsExpressionContextUtils::projectScope( QgsProject::instance() );
```
Comment (nyalldawson) [inline src/3d/qgsfeature3dhandler_p.cpp]: We need a "3d map settings" scope... but I can't think of anything to put in it yet :wink:

File: src/3d/qgsfeature3dhandler_p.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    exprContext.setFields( layer->fields() );
+    context.setExpressionContext( exprContext );
+
+    QSet<QString> attributeNames;
+    if ( !handler->prepare( context, attributeNames ) )
+      return nullptr;
+
+    // build the feature request
+    QgsFeatureRequest req;
+    req.setDestinationCrs( map.crs(), map.transformContext() );
```
Comment (nyalldawson) [inline src/3d/qgsfeature3dhandler_p.cpp]: Do 3d views ALWAYS use the full extent of layers?

File: src/3d/qgsrulebased3drenderer.cpp [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+  else
+  {
+    mFilter.reset( nullptr );
+  }
+}
+
+void QgsRuleBased3DRenderer::Rule::updateElseRules()
+{
+  mElseRules.clear();
+  Q_FOREACH ( Rule *rule, mChildren )
```
Comment (nyalldawson) [inline src/3d/qgsrulebased3drenderer.cpp]: Q_FOREACH? waaaaaat? :stuck_out_tongue:

File: src/3d/qgsrulebased3drenderer.cpp [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+    // add handler!
+    Q_ASSERT( !handlers.value( this ) );
+    QgsFeature3DHandler *handler = nullptr;
+    if ( mSymbol->type() == QLatin1String( "polygon" ) )
+    {
+      handler = Qgs3DSymbolImpl::handlerForPolygon3DSymbol( layer, *static_cast<QgsPolygon3DSymbol *>( mSymbol.get() ) );
+    }
+    else if ( mSymbol->type() == QLatin1String( "point" ) )
+    {
+      // TODO: point symbol
```
Comment (nyalldawson) [inline src/3d/qgsrulebased3drenderer.cpp]: is this still a todo?

File: src/3d/qgsrulebased3drenderer.cpp [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+  RuleToHandlerMap handlers;
+  mRootRule->createHandlers( vl, handlers );
+
+  QSet<QString> attributeNames;
+  mRootRule->prepare( context, attributeNames, handlers );
+
+  QgsFeatureRequest req;
+  req.setDestinationCrs( map.crs(), map.transformContext() );
+  req.setSubsetOfAttributes( attributeNames, vl->fields() );
```
Comment (nyalldawson) [inline src/3d/qgsrulebased3drenderer.cpp]: I noticed you haven't ported the equivalent of QgsRuleBasedRenderer::filter for setting an expression based filter on the request based on the visible rules -- this is a huge performance boost for rule based renderer, and would probably help here too (and is likely just more copy+past from rule based renderer)

File: src/3d/qgsrulebased3drenderer.cpp [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+
+  QgsFeature f;
+  QgsFeatureIterator fi = vl->getFeatures( req );
+  while ( fi.nextFeature( f ) )
+  {
+    // TODO: set feature in expr context?
```
Comment (nyalldawson) [inline src/3d/qgsrulebased3drenderer.cpp]: yep

File: src/3d/qgsrulebased3drenderer.cpp [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+
+Qt3DCore::QEntity *QgsRuleBased3DRenderer::createEntity( const Qgs3DMapSettings &map ) const
+{
+  QgsVectorLayer *vl = layer();
+
+  if ( !vl )
+    return nullptr;
+
+  Qgs3DRenderContext context( map );
+  // TODO: add expr. context
```
Comment (nyalldawson) [inline src/3d/qgsrulebased3drenderer.cpp]: this looks to be partly done at least?

File: src/app/3d/qgsrulebased3drendererwidget.cpp [hunk 10]
```diff
@@ -0,0 +1,654 @@
...
+
+Qgs3DRendererRulePropsWidget::~Qgs3DRendererRulePropsWidget()
+{
+  delete mSymbol;
+}
+
+void Qgs3DRendererRulePropsWidget::setDockMode( bool dockMode )
+{
+  QgsPanelWidget::setDockMode( dockMode );
+  // TODO mLabelingGui->setDockMode( dockMode );
```
Comment (nyalldawson) [inline src/app/3d/qgsrulebased3drendererwidget.cpp]: not needed?

**Comment by nirvn** [comment]
Super exciting stuff @wonder-sk !

**Comment by nirvn** [comment]
Btw I'd use the 2d simple & rule based renderer icons in the combo box here.

File: src/3d/qgsfeature3dhandler_p.cpp (hunk 1 above)
Comment (wonder-sk) [reply src/3d/qgsfeature3dhandler_p.cpp]: probably yes - unified expr context creation in Qgs3DUtils now

File: src/3d/qgsfeature3dhandler_p.cpp (hunk 3 above)
Comment (wonder-sk) [reply src/3d/qgsfeature3dhandler_p.cpp]: At some point yes :)

File: src/3d/qgsfeature3dhandler_p.cpp (hunk 4 above)
Comment (wonder-sk) [reply src/3d/qgsfeature3dhandler_p.cpp]: At this point yes... only terrain currently uses tiling (but it has been designed in a way that other 3D entities may use it in the future too)

File: src/3d/qgsrulebased3drenderer.cpp (hunk 6 above)
Comment (wonder-sk) [reply src/3d/qgsrulebased3drenderer.cpp]: Yes - going to address it in a bit

File: src/3d/qgsrulebased3drenderer.cpp (hunk 5 above)
Comment (wonder-sk) [reply src/3d/qgsrulebased3drenderer.cpp]: a small trap to check if reviewers are paying attention ;-) fixed in a01caa43

File: src/app/3d/qgsrulebased3drendererwidget.cpp [hunk 11]
```diff
@@ -0,0 +1,632 @@
...
+#include "qgsrulebased3drendererwidget.h"
+
+#include "qgs3dutils.h"
+#include "qgsexpressionbuilderdialog.h"
+#include "qgslogger.h"
+#include "qgsrulebased3drenderer.h"
+#include "qgsvectorlayer.h"
+#include "qgssymbol3dwidget.h"
+
+#include <QClipboard>
```
[the remark this answers is not in the document]
Comment (wonder-sk) [reply src/app/3d/qgsrulebased3drendererwidget.cpp]: fixed in d1cc1ee

[the remark this answers is not in the document]
Comment (nirvn) [reply src/app/3d/qgsrulebased3drendererwidget.cpp]: That's a very personalized commit string :wink:

**Comment by nirvn** [comment]
@wonder-sk , it'd be nice if the combo box margin could align with that of [2d] symbology and labelling, check this GIF:
peek 2019-01-18 17-41

**Comment by wonder-sk** [comment]
@nirvn margin fixed in de5c7cb

**Comment by nirvn** [comment]
No other comments from me; this is a very useful feature.

File: src/3d/qgsrulebased3drenderer.cpp (hunk 7 above)
Comment (wonder-sk) [reply src/3d/qgsrulebased3drenderer.cpp]: I have based the implementation on rule-based labeling which does not have it... yeah I can add support for this but would leave it up for a later time if you don't mind (preparation time here is typically dominated by tessellation)

**Comment by wonder-sk** [comment]
OK Let me merge this to meet feature freeze deadline and sort out point 3D symbols afterwards...

**Pull request merged**